File: 2018/burger/experimental/dek/burgermaker/burger_checker.py
import sys
sys.path.insert(0, "../burgerdata")
from burger_data import BurgerElement

def check_burger(burger, debug=False):
  if len(burger) != 8:
    if debug: print "Burger is wrong size"
    return False
  crown_pos = None
  for i in range(len(burger)):
    layer = burger[i]
    # base case
    if layer == BurgerElement.empty:
      continue
    if layer == BurgerElement.crown:
      crown_pos = i
      break
    if debug: print "prefix elements must be empty or crown"
    return False
  if crown_pos is None:
    if debug: print "Burger does not have crown"
    return False
  if burger[-1] != BurgerElement.heel:
    if debug: print "Burger does not have heel bottom"
    return False
  for i in range(crown_pos+1, len(burger)-1):
    if burger[i] == BurgerElement.crown or burger[i] == BurgerElement.heel:
      if debug: print "Cannot have internal buns"
      return False
  for i in range(crown_pos+1, len(burger)):
    if burger[i] == BurgerElement.empty:
      if debug: print "Cannot have internal or terminal empty"
      return False
  for i in range(len(burger)-1):
    first, second = burger[i], burger[i+1]
    if first != BurgerElement.empty and second != BurgerElement.empty and first == second:
      if debug: print "Cannot have identical sequential items"
      return False
  for i in range(1, len(burger)-1):
    if burger[i] == BurgerElement.patty:
      break
  else:
    if debug: print "Must have at least one patty"
    return False
  last_cheese_index = None
  for i in range(1, len(burger)-1):
    if burger[i] == BurgerElement.cheese:
      last_cheese_index = i
  if last_cheese_index is not None:
    for i in range(last_cheese_index+1, len(burger)):
      if burger[i] == BurgerElement.patty:
        break
    else:
      if debug: print "Must have at patty under last cheese."
      return False


  return True

if __name__ == '__main__':
  test_burgers = [
    [BurgerElement.crown, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.cheese, BurgerElement.patty, BurgerElement.cheese, BurgerElement.patty, BurgerElement.heel],
    [BurgerElement.crown, BurgerElement.empty, BurgerElement.tomato, BurgerElement.cheese, BurgerElement.patty, BurgerElement.cheese, BurgerElement.patty, BurgerElement.heel],
    [BurgerElement.tomato, BurgerElement.crown, BurgerElement.tomato, BurgerElement.cheese, BurgerElement.patty, BurgerElement.cheese, BurgerElement.patty, BurgerElement.heel],
    [BurgerElement.crown, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.cheese, BurgerElement.patty, BurgerElement.cheese, BurgerElement.patty, BurgerElement.tomato],
    [BurgerElement.crown, BurgerElement.tomato, BurgerElement.tomato, BurgerElement.cheese, BurgerElement.patty, BurgerElement.cheese, BurgerElement.patty, BurgerElement.heel],
    [BurgerElement.crown, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.heel],
    [BurgerElement.crown, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.lettuce, BurgerElement.tomato, BurgerElement.patty, BurgerElement.cheese, BurgerElement.heel],
    [BurgerElement.heel, BurgerElement.patty, BurgerElement.heel, BurgerElement.patty, BurgerElement.heel, BurgerElement.patty, BurgerElement.crown, BurgerElement.heel]
  ]

  for test_burger in test_burgers:
    print test_burger
    print check_burger(test_burger, debug=True)

File: 2018/burger/experimental/dek/burgermaker/burger_checker.cc
#include <msgpack.hpp>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>

using namespace std;

enum BurgerElements {
  EMPTY = '0',
  CROWN,
  LETTUCE,
  TOMATO,
  CHEESE,
  PATTY,
  HEEL,
};
  
bool check_burger(string burger, bool debug = false) {
  if (burger.length() != 8) {
    if (debug) cout << "Burger is wrong size" << endl;
    return false;
  }
  string::iterator crown_pos = burger.end();
  for (auto layer_it = burger.begin(); layer_it != burger.end(); ++layer_it) {
    if (*layer_it == EMPTY) continue;
    if (*layer_it == CROWN) {
      crown_pos = layer_it;
      break;
    }
    if (debug) cout << "Prefix elements must be empty or crown" << endl;
    return false;
  }
  
  if (crown_pos == burger.end()) {
    if (debug) cout << "Must have crown." << endl;
    return false;
  }

  if (burger[7] != HEEL) {
    if (debug) cout << "Burger must have heel bottom" << endl;
    return false;
  }

  if ((find(crown_pos+1, burger.end()-1, CROWN) != burger.end()-1) ||
      (find(crown_pos+1, burger.end()-1, HEEL) != burger.end()-1)) {
    if (debug) cout << "Cannot have internal buns" << endl;
    return false;
  }

  if ((find(crown_pos+1, burger.end(), EMPTY) != burger.end())) {
    if (debug) cout << "Cannot have internal or terminal empty" << endl;
    return false;
  }

  char prev = '\0';
  for(auto &it : burger) {
    if (it != '0' && it == prev) {
      if (debug) cout << "Cannot have sequential non-empty items" << endl;
      return  false;
    }
    prev = it;
  }

  if (burger.find(PATTY) == string::npos) {
    if (debug) cout << "Must have at least one patty." << endl;
    return false;
  }

  auto last_cheese_pos = burger.find_last_of(CHEESE);
  if (last_cheese_pos != string::npos) {
    auto last_patty_pos = burger.find_last_of(PATTY);
    if (last_patty_pos != string::npos && last_patty_pos < last_cheese_pos) {
      if (debug) cout << "Must have at least one patty after last cheese." << endl;
      return false;
    }
  }

  return true;
}

int main() {

  ifstream file("burgers.txt");
  string row_index;
  string burger;
  cout << boolalpha;
  while(file.good()) {
    getline(file, burger, '\n') ;
    if (!file.good()) break;
    cout << burger << " " << check_burger(burger) << endl;
  }

  // burger = "10000156";
  // cout << burger << endl;
  // check_burger(burger, true);
}
